tests/cycle_output.rs

```diff
@@ -163,16 +163,15 @@ fn revalidate_with_change_after_output_read() {
 
     db.assert_logs(expect![[r#"
         [
-            "salsa_event(DidValidateMemoizedValue { database_key: read_value(Id(401)) })",
+            "salsa_event(DidValidateMemoizedValue { database_key: read_value(Id(403)) })",
             "salsa_event(WillExecute { database_key: query_d(Id(800)) })",
             "salsa_event(WillExecute { database_key: query_a(Id(0)) })",
-            "salsa_event(WillExecute { database_key: read_value(Id(400)) })",
+            "salsa_event(DidValidateMemoizedValue { database_key: read_value(Id(400)) })",
```

Comment:
Because salsa now keeps the intermediate tracked struct around, it can now also reuse the results on those queries. That means, this fix also improves incremental caching in cycles across revisions.

Reply:
Okay, I was a bit too optimistic. I think this optimization could lead to a memory leak :( I'll write a more detailed comment on the relevant line